### DISCUSSION TITLE
Libvirt backend support

### DISCUSSION THREAD

**Discussion by clarfonthey** [woodpecker-ci/woodpecker#2302] [Request a Feature]
### Clear and concise description of the problem

Docker containers offer a majority of the cases needed for builds, but in the case of non-Linux systems like OpenBSD, NetBSD, etc., launching complete virtual machines for builds may be required. We need a new backend to support this.

### Suggested solution

Libvirt is probably the most versatile system for doing this, since it provides a common API for running things under several different hypervisors.

### Alternative

Without libvirt, we'd probably have to manage multiple hypervisor platforms individually. However, it does require a separate runtime layer, meaning that it does add an extra dependency. If we were to implement these hypervisor backends directly, it wouldn't make libvirt redundant completely, but it would make the need for it less pronounced.

### Additional context

Most commercial CI systems allow launching full VMs for their versatility, although it's obviously much slower than using containers. To fully replace these providers, woodpecker would need a VM-based runtime.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by 6543** [comment]
the biggest question would be, how to handle "images" ... - is there a common way / index for vm images ?!?

**Comment by danog** [comment]
About this, just found https://github.com/containers/krunvm which can create actual VMs using KVM from plain OCI images!

**Comment by namgo** [comment]
> index for vm images

Libvirt provides no globally available registry but there are storage-pools. https://libvirt.org/storage.html#network-filesystem-pool suggests CIFS/NFS/glusterfs as principally supported network protocols for accessing storage pools, the expectation is a LAN environment.

Neither vagrant nor packer provide a container registry feature. Network-booting a live image might work in some scenarios but I don't feel as though it is standardized enough - hoping I'm wrong.

A hosting organization's servers might provide available golden images to users like you might see in a container registry.

I would expect to provide a series of pools on each libvirt-runner. I would assume "`image_libvirt:`" would expect "`pool`/`image_name`" or simply enumerate pools and select `image_name`.

Edit: I re-organized thoughts for clarity.

I apologize since I'm new to this repo and don't know all the norms, just checking it out trying to see if it fits my use-case.

Personally I've been looking for a CI/CD system for the Plan9Front Operating System, I suspect the Local backend will be mostly fine for my needs but it will require me to spin up dedicated Plan9 infrastructure; libvirt support would be of great convenience to me!

**Comment by computerquip-work** [comment]
I'm interested in this as well. I'd expect it to either use images directly or use something like vagrant boxes. In my experience, most images are just hosted through a basic file server whether that be HTTP, FTP, or whatever. Honestly, I'd be happy with allowing a remote url to download an image.

This would be particularly useful for Windows where you can't run containers for Windows on Linux. In addition, this is still useful for things like testing drivers and installers which containers aren't fit for generally.

**Comment by 6543** [comment]
next to make freebsd support easyer.... do priviledged actions like FUSE mounts securly would also require a VM backend

-> https://codeberg.org/Codeberg-CI/feedback/issues/239#issuecomment-9728450

**Comment by mdekstrand** [comment]
Even without support for an image registry, I would find it useful to have a backend that works like like the local backend, except that each pipeline is run in a fresh VM with libvirt — the base image could even be statically configured in the agent configuration for such a design.

**Comment by sirikon** [comment]
For anyone interested, if this PR lands (https://github.com/woodpecker-ci/woodpecker/pull/6150), it should be possible to glue something together to create a VM, run an ephemeral agent on it, wait for the process to end, kill the VM and repeat.

**Reply by mdekstrand** [reply]
I like this solution / idea, and perhaps libvirt support would be better for the autoscaler than for the agent itself.

In particular, there are two distinct configuration goals I see for `libvirt` (or similar) support, and separating “run the VM” from “run the agent on the VM” allows both to be supported:

- A `local` agent with full access, running on the VM. Useful for things that don't work well inside Docker containers but still want isolation. For example, I've been looking at using Woodpecker to automate building `bootc` images for deployment on my homelab, and building the bootc image from inside Docker doesn't work very well. (For this private-only case, I'm planning to just use a local agent without isolation, but the isolation would be a nice add-on.)
- Normal Docker-based pipelines, but where the runner wants VM-level isolation instead of just Docker-level isolation (e.g., for untrusted code). For this use case, spinning up a bare VM image with Docker and an ephemeral agent with the Docker backend would fit the bill well (modulo some additional caching infrastructure).

If the autoscaler (or a similar component) knew how to spin up the `libvirt` VM with an appropriate local or Docker ephemeral agent, then both can be supported with one code path just by how you set up the VM image.

**Reply by 6543** [reply]
https://github.com/woodpecker-ci/autoscaler/issues/522 that would be :)

**Reply by hasufell** [reply]
@sirikon why an ephemeral agent? Don't we want a "static" agent to spawn an ephemeral VM? At least that seems to be the case when those VMs are to be spawned on self-hosted runners.

**Reply by mdekstrand** [reply]
@hasufell Having the agent know how to spawn the VM means that the agent would need to know how to spawn VMs. This could be a useful backend (would it work like the `"local"` backend, except inside a VM?). However, since the auto-scaler or other external orchestration already knows how to spawn VMs, using an ephemeral agent allows an implementation shortcut where the “how to spawn VMs” knowledge only exists in the auto-scaler or orchestration code.

It does leave a hole where a "local" ephemeral agent inside the VM is still not usable for untrusted code, because it has access to the woodpecker shared secret.

**Reply by 6543** [reply]
the autoscaler dont use shared secrets :tada:

**Reply by mdekstrand** [reply]
@6543 That is very good to know! Makes it even better :).

**Reply by mdekstrand** [reply]
Just for the record / completeness, `libkrun` (which is used by crun-krun) may be a better or easier solution than targeting libvirt, although both are likely useful in the autoscaler. A `crun-krun` solution could also help solve the image problem — run Docker images with krun isolation.

**Comment by hasufell** [comment]
https://github.com/woodpecker-ci/woodpecker/pull/6985

**Reply by hasufell** [reply]
Anyone willing to test this?
